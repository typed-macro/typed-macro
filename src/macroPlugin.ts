import type { Plugin, ViteDevServer } from 'vite'
import { NormalizedExports } from '@/core/exports'
import { MacroProvider } from '@/macroProvider'
import { DevServerHelper, getDevServerHelper } from '@/helper/server'
import { Runtime, RuntimeOptions } from '@/core/runtime'

export type MacroPluginHooks = Omit<
  Plugin,
  'name' | 'enforce' | 'apply' | 'configureServer'
> & {
  configureServer?: (
    server: ViteDevServer,
    helpers: DevServerHelper
  ) => (() => void) | void | Promise<(() => void) | void>
}

export type InternalPluginOptions = {
  name: string
  runtimeOptions: RuntimeOptions
  exports: NormalizedExports
  hooks: MacroPluginHooks
}

export interface MacroPlugin extends Plugin {
  __consume: () => MacroProvider
}

interface InternalMacroPlugin extends MacroPlugin {
  __internal_macro_plugin: true
}

export function macroPlugin(options: InternalPluginOptions): MacroPlugin {
  const {
    name,
    exports,
    runtimeOptions,
    hooks: {
      buildStart,
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    },
  } = options

  let runtime: Runtime | undefined = new Runtime(runtimeOptions)

  runtime.register(exports)

  return {
    __internal_macro_plugin: true,
    __consume() {
      if (!runtime) throw new Error(`plugin '${name}' is used more than once.`)
      const provider: MacroProvider = {
        id: name,
        exports: runtime.exports,
        hooks: {},
      }
      runtime = undefined
      return provider
    },
    name,
    enforce: 'pre',
    async buildStart(opt) {
      await Promise.all([
        runtime?.typeRenderer.write(),
        buildStart?.bind(this)(opt),
      ])
    },
    configResolved(config) {
      runtime?.setDevMode(config?.env?.DEV ?? false)
      return configResolved?.(config)
    },
    resolveId(id, importer, options, ssr) {
      const result = runtime?.handleResolveId(id)
      if (result) return result
      return resolveId?.bind(this)(id, importer, options, ssr)
    },
    load(id, ssr) {
      const result = runtime?.handleLoad(id)
      if (result) return result
      return load?.bind(this)(id, ssr)
    },
    transform(code, id, ssr) {
      const transformed = runtime?.handleTransform(code, id, ssr)
      return transform?.bind(this)(transformed ?? code, id, ssr) ?? transformed
    },
    configureServer(server) {
      // hook
      if (configureServer) configureServer(server, getDevServerHelper(server))
    },
    ...otherHooks,
  } as InternalMacroPlugin
}

export function isMacroPlugin(o: unknown): o is MacroPlugin {
  return (o as any).__internal_macro_plugin ?? false
}
