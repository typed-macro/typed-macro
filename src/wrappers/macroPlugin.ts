import type { Plugin, ViteDevServer } from 'vite'
import { DevServerHelper, getDevServerHelper } from '@/wrappers/helper/server'
import { Attachable, Runtime } from '@/core/runtime'
import { versionedPlugin } from '@/wrappers/compat'

export type MacroPluginHooks = Omit<
  Plugin,
  'name' | 'enforce' | 'apply' | 'configureServer'
> & {
  configureServer?: (
    server: ViteDevServer,
    helpers: DevServerHelper
  ) => (() => void) | void | Promise<(() => void) | void>
}

export type InternalPluginOptions = {
  name: string
  runtime: Runtime
  hooks: MacroPluginHooks
}

export interface MacroPlugin extends Plugin {
  __consume: () => Attachable
}

interface InternalMacroPlugin extends MacroPlugin {
  __internal_macro_plugin: true
}

export function macroPlugin(options: InternalPluginOptions): MacroPlugin {
  const {
    name,
    hooks: {
      buildStart,
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    },
  } = options

  let runtime: Runtime | undefined = options.runtime

  const plugin = {
    __internal_macro_plugin: true,
    __consume() {
      if (!runtime) throw new Error(`plugin '${name}' is used more than once.`)
      const consume = {
        exports: runtime.exports,
        options: runtime.options,
      }
      runtime = undefined
      return consume
    },
    name,
    enforce: 'pre',
    configResolved(config) {
      runtime?.setDevMode(config?.env?.DEV ?? false)
      return configResolved?.(config)
    },
    configureServer(server) {
      if (configureServer) configureServer(server, getDevServerHelper(server))
    },
    async buildStart(opt) {
      await Promise.all([
        runtime?.typeRenderer.write(),
        buildStart?.call(this, opt),
      ])
    },
    resolveId(id, importer, options, ssr) {
      const result = runtime?.handleResolveId(id)
      if (result) return result
      return resolveId?.call(this, id, importer, options, ssr)
    },
    load(id, ssr) {
      const result = runtime?.handleLoad(id)
      if (result) return result
      return load?.call(this, id, ssr)
    },
    transform(code, id, ssr) {
      const transformed = runtime?.handleTransform(code, id, ssr)
      return (
        transform?.call(this, transformed ?? code, id, ssr) ??
        (transformed && { code: transformed, map: null })
      )
    },
    ...otherHooks,
  } as InternalMacroPlugin
  return versionedPlugin(plugin)
}

export function isMacroPlugin(o: unknown): o is MacroPlugin {
  return (o as any)?.__internal_macro_plugin ?? false
}
