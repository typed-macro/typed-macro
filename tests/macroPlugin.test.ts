import { isMacroPlugin, MacroPlugin, macroPlugin } from '@/macroPlugin'
import { withTempPath } from './testutils'

describe('macroPlugin() & isMacroPlugin()', () => {
  it('should work', () => {
    expect(
      isMacroPlugin(
        macroPlugin({
          exports: {
            types: {},
            modules: {},
            macros: {},
          },
          hooks: {},
          name: 'test',
          runtimeOptions: {
            typeRenderer: {
              typesPath: '',
            },
            transformer: {},
          },
        })
      )
    ).toBe(true)
  })
})

describe('MacroPlugin', () => {
  let _plugin: MacroPlugin
  let mockedHook: jest.Mock
  beforeEach(() => {
    mockedHook = jest.fn()
    _plugin = macroPlugin({
      name: 'test',
      exports: {
        modules: {
          '@helper': `export function hello(msg) {console.log(msg)}`,
        },
        macros: {
          '@echo': [
            {
              name: 'echo',
              apply: ({ path }, { template }, { prependToBody }) => {
                prependToBody(
                  template.statements.ast(
                    `import { hello } from '@helper'; hello('world')`
                  )
                )
                path.remove()
              },
            },
          ],
        },
        types: {},
      },
      hooks: {
        transform: mockedHook,
        resolveId: mockedHook,
        load: mockedHook,
        configureServer: mockedHook,
        configResolved: mockedHook,
        buildStart: mockedHook,
      },
      runtimeOptions: {
        transformer: {},
        typeRenderer: { typesPath: '' },
      },
    })
  })

  it('should support to be consumed', () => {
    expect(_plugin.__consume()).toMatchSnapshot()
    // consume more than once
    expect(() => _plugin.__consume()).toThrow()
  })

  it('should handle transform with macro', () => {
    expect(
      _plugin.transform!.call(null as any, `import { echo } from '@echo'`, '')
    ).toBeUndefined()
    expect(
      _plugin.transform!.call(
        null as any,
        `import { echo } from '@echo'; echo()`,
        'a.ts'
      )
    ).toMatchSnapshot()
    expect(mockedHook.mock.calls.length).toBe(2)
  })

  it('should stop handling transform after consumed', () => {
    _plugin.__consume()
    expect(
      _plugin.transform!.call(
        null as any,
        `import { echo } from '@echo'; echo()`,
        'a.ts'
      )
    ).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(1)
  })

  it('should handle load properly', () => {
    expect(_plugin.load!.call(null as any, '@helper')).toMatchSnapshot()
    expect(_plugin.load!.call(null as any, '@echo')).toMatchSnapshot()

    expect(_plugin.load!.call(null as any, '@others')).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(1)
  })

  it('should stop handling load after consumed', () => {
    _plugin.__consume()
    expect(_plugin.load!.call(null as any, '@helper')).toBeUndefined()
    expect(_plugin.load!.call(null as any, '@echo')).toBeUndefined()
    expect(_plugin.load!.call(null as any, '@others')).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(3)
  })

  it('should handle resolveId properly', () => {
    expect(_plugin.resolveId!.call(null as any, '@helper', '', {})).toBe(
      '@helper'
    )
    expect(_plugin.resolveId!.call(null as any, '@echo', '', {})).toBe('@echo')

    expect(
      _plugin.resolveId!.call(null as any, '@others', '', {})
    ).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(1)
  })

  it('should stop handling resolveId after consumed', () => {
    _plugin.__consume()
    expect(
      _plugin.resolveId!.call(null as any, '@helper', '', {})
    ).toBeUndefined()
    expect(
      _plugin.resolveId!.call(null as any, '@echo', '', {})
    ).toBeUndefined()
    expect(
      _plugin.resolveId!.call(null as any, '@others', '', {})
    ).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(3)
  })

  it('should handle buildStart properly', (done) => {
    withTempPath('a.d.ts', async (tempPath) => {
      _plugin = macroPlugin({
        name: 'test',
        exports: {
          modules: {},
          macros: {},
          types: {},
        },
        hooks: { buildStart: mockedHook },
        runtimeOptions: {
          transformer: {},
          typeRenderer: { typesPath: tempPath },
        },
      })
      await _plugin.buildStart!.call(null as any, {} as any)
      expect(mockedHook.mock.calls.length).toBe(1)
      done()
    })
  })

  it('should handle configResolved properly', async () => {
    await _plugin.configResolved!.call(null as any, {} as any)
    expect(mockedHook.mock.calls.length).toBe(1)
  })
})
