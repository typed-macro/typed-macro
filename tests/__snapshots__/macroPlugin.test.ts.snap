// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MacroPlugin should handle load properly 1`] = `"export function hello(msg) {console.log(msg)}"`;

exports[`MacroPlugin should handle load properly 2`] = `"export {}"`;

exports[`MacroPlugin should handle transform with macro 1`] = `"import { hello } from '@helper';hello('world');"`;

exports[`MacroPlugin should support to be consumed 1`] = `
Object {
  "exports": Object {
    "macros": Object {
      "@echo": Array [
        Object {
          "apply": [Function],
          "name": "echo",
        },
      ],
    },
    "modules": Object {
      "@helper": "export function hello(msg) {console.log(msg)}",
    },
    "types": Object {},
  },
  "hooks": Object {},
  "id": "test",
}
`;
